experiment_scripts/rosenbrock.py

```diff
@@ -1,57 +1,188 @@
-import os
+import time
 
+import matplotlib.pyplot as plt
 import torch
-from botorch.acquisition import qKnowledgeGradient
+from botorch.acquisition import (
+    qKnowledgeGradient,
+    HybridKnowledgeGradient,
+    DiscreteKnowledgeGradient,
+    ContinuousKnowledgeGradient,
+)
 from botorch.fit import fit_gpytorch_model
+from botorch.generation import get_best_candidates, gen_candidates_torch
 from botorch.models import SingleTaskGP
+from botorch.optim import gen_batch_initial_conditions
 from botorch.optim import optimize_acqf
 from botorch.test_functions import Rosenbrock
 from botorch.utils import standardize
 from botorch.utils.sampling import manual_seed
 from botorch.utils.transforms import unnormalize
 from gpytorch.mlls import ExactMarginalLogLikelihood
 
-SMOKE_TEST = os.environ.get("SMOKE_TEST")
-
-# Defining box optimisation bounds
-bounds = torch.stack([torch.zeros(2), torch.ones(2)])
-
-# Generate XY data
-train_X = bounds[0] + (bounds[1] - bounds[0]) * torch.rand(20, 2)
 fun = Rosenbrock(dim=2, negate=True)
 fun.bounds[0, :].fill_(-5)
 fun.bounds[1, :].fill_(10)
 
+# Defining box optimisation bounds
+bounds = torch.stack([torch.zeros(2), torch.ones(2)])
+bounds[:, 0].fill_(-5)
+bounds[:, 1].fill_(10)
+
 
 def eval_objective(x):
     """This is a helper function we use to unnormalize and evaluate a point"""
     return fun(unnormalize(x, fun.bounds)).unsqueeze(-1)
 
 
+# Generate XY data
+torch.manual_seed(1)
+train_X = bounds[:, 0] + (bounds[:, 1] - bounds[:, 0]) * torch.rand(30, 2)
 train_Y = standardize(eval_objective(train_X))
 
-
 # Fit GP model
 model = SingleTaskGP(train_X, train_Y)
 mll = ExactMarginalLogLikelihood(model.likelihood, model)
 fit_gpytorch_model(mll)
 
-# Defining the qKnowledgeGradient acquisition function (One-Shot KG)
-NUM_FANTASIES = 128
-qKG = qKnowledgeGradient(model, num_fantasies=NUM_FANTASIES)
+# acquisition function and optimisation parameters
+NUM_RESTARTS = 5
+RAW_SAMPLES = 50
+
+# Initialize acquisition functions.
+NUM_FANTASIES_ONE_SHOT = 125
+one_shot_kg = qKnowledgeGradient(model, num_fantasies=NUM_FANTASIES_ONE_SHOT)
+
+NUM_DISCRETE_X = 100
+discrete_kg = DiscreteKnowledgeGradient(
+    model=model,
+    bounds=fun.bounds,
+    num_discrete_points=NUM_DISCRETE_X,
+    discretisation=None,
+)
+
+NUM_FANTASIES_CONTINUOUS_KG = 5
+continous_kg = ContinuousKnowledgeGradient(
+    model,
+    bounds=fun.bounds,
+    num_fantasies=NUM_FANTASIES_CONTINUOUS_KG,
+    num_restarts=1,
+    raw_samples=20,
+)
+
+
+NUM_FANTASIES_HYBRID_KG = 5
+hybrid_kg = HybridKnowledgeGradient(
+    model=model,
+    bounds=fun.bounds,
+    num_fantasies=NUM_FANTASIES_HYBRID_KG,
+    num_restarts=1,
+    raw_samples=20,
+)
+
+# Optimise acquisition functions
+with manual_seed(12):
 
-# Optimize acquisition function
-NUM_RESTARTS = 10
-RAW_SAMPLES = 512
+    # Hybrid KG optimisation with Adam's optimiser
+    start = time.time()
+    hybrid_kg_xstar, _ = optimize_acqf(
```

Comment:
rename variable `x_star_hkg`? (it is a point in X space)

Reply:
done